### PR TITLE
pkg: finder -> resourcemonitor

### PR DESCRIPTION
**Pull request by ffromani**
This PR drops the existing pkg/finder and replaces with
pkt/resourcemonitor. The functionality of the package is preserved with
minimal API change, and no change of behaviour is intended for
`nfd-topology-updater`.

Rationale for this rewrite is that `pkg/finder` mostly grow up over
time with no organic design and as aggregation of feature.
Thus, the design was messy and the code was hard to follow and to test.

The current rewrite wants to preserve the responsabilities and
the basic design tenets of the package, but aims to provide
a much cleaner interface, better readability and improved testability.

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

File: pkg/topologyinfo/cpus/cpus.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * Copyright 2020 Red Hat, Inc.
+ */
+
+package cpus
+
+import (
+	"strconv"
+
+	"github.com/fromanirh/topologyinfo/sysfs"
```
Comment (ffromani) [inline pkg/topologyinfo/cpus/cpus.go]: wrong import, will fix in the next upload.

**Comment by cynepco3hahue** [comment]
@fromanirh Does it WIP? Because I can not see the implementation for
```
nodes, err := numa.NewNodesFromSysFS(sysfsPath)
if err != nil {
    return nil, err
}

dists, err := distances.NewDistancesFromSysfs(sysfsPath)
if err != nil {
	return nil, err
}
```

**Comment by ffromani** [comment]
> @fromanirh Does it WIP? Because I can not see the implementation for
> 
> ```
> nodes, err := numa.NewNodesFromSysFS(sysfsPath)
> if err != nil {
>     return nil, err
> }
> 
> dists, err := distances.NewDistancesFromSysfs(sysfsPath)
> if err != nil {
> 	return nil, err
> }
> ```

Already merged on a previous PR: https://github.com/swatisehgal/node-feature-discovery/tree/rte/pkg/topologyinfo/numa

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	return zones
+}
+
+func GetContainerDevicesFromAllocatableResources(availRes *podresourcesapi.AvailableResourcesResponse, cpus *cpus.CPUs) []*podresourcesapi.ContainerDevices {
+	var contDevs []*podresourcesapi.ContainerDevices
+	for _, dev := range availRes.GetDevices() {
+		contDevs = append(contDevs, dev)
+	}
+
+	cpuRes := make(map[int][]string)
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go]: I think this should be named cpuResPerNuma or on those lines to be more descriptive

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+				Nodes: []*podresourcesapi.NUMANode{
+					{ID: int64(nodeID)},
+				},
+			},
+		})
+	}
+
+	return contDevs
+}
+
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go:181]: Lets add a comment providing a brief description of what this function does

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		numaData[nodeNum][ri.Name].allocatable--
+	}
+}
+
+func makeZoneName(numaID int) string {
+	return fmt.Sprintf("node-%d", numaID)
+}
+
+func MakeNodeCapacity(devices []*podresourcesapi.ContainerDevices) map[int]map[v1.ResourceName]int64 {
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go:209]: Same as above

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+			}
+			nodeRes[v1.ResourceName(resourceName)] = int64(len(device.GetDeviceIds()))
+			perNUMACapacity[nodeNum] = nodeRes
+		}
+	}
+	return perNUMACapacity
+
+}
+
+func makeResourceMap(numaNodes int, devices []*podresourcesapi.ContainerDevices) map[string]map[string]int {
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go:230]: Same

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+			for _, deviceID := range device.GetDeviceIds() {
+				deviceID2NUMAID[deviceID] = nodeNuma
+			}
+		}
+		resourceMap[device.GetResourceName()] = deviceID2NUMAID
+	}
+	return resourceMap
+}
+
+func makeCostsPerNumaNode(numaInfo *numa.Nodes, numaDists *distances.Distances, numaIDSrc int) (map[string]int, error) {
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go:247]: Here too

File: pkg/topologyinfo/cpus/cpus.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+package cpus
+
+import (
+	"strconv"
+
+	"sigs.k8s.io/node-feature-discovery/pkg/topologyinfo/sysfs"
+)
+
+// CPUIdList is a list of CPU IDs (integer core identifier)
```
Comment (swatisehgal) [inline pkg/topologyinfo/cpus/cpus.go]: This is great, it would be great to have description like this for other data structures in type.go too

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+	for nodeNum, resList := range perNuma {
+		zone := &v1alpha1.Zone{
+			Type:      "Node",
+			Resources: make(v1alpha1.ResourceInfoMap),
+		}
+
+		costs, err := makeCostsPerNumaNode(nodeResourceData.numaInfo, nodeResourceData.numaDists, nodeNum)
+		if err != nil {
+			// TODO log and move forward
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go]: Lets add a log message here

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+	var contDevs []*podresourcesapi.ContainerDevices
+	for _, dev := range availRes.GetDevices() {
+		contDevs = append(contDevs, dev)
+	}
+
+	cpuRes := make(map[int][]string)
+	for _, cpuID := range availRes.GetCpuIds() {
+		nodeID, ok := cpus.GetNodeIDForCPU(int(cpuID))
+		if !ok {
+			// TODO log this is weird
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go]: Lets add a log here

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+		cpuInfo:           cpuInfo,
+		numaInfo:          &nodes,
+		numaDists:         dists,
+		resourceID2NUMAID: makeResourceMap(len(nodes.Online), allDevs),
+		perNUMACapacity:   MakeNodeCapacity(allDevs),
+	}
+
+	return &nodeResourceInstance, nil
+}
+
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go:94]: Lets add a comment providing a brief description of what this function does

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+			zone.Resources[name.String()] = v1alpha1.ResourceInfo{
+				Allocatable: allocatableQty.String(),
+				Capacity:    capacityQty.String(),
+			}
+		}
+		zones[makeZoneName(nodeNum)] = zone
+	}
+	return zones
+}
+
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go:143]: Lets add a comment providing a brief description of what this function does

**Review by swatisehgal — CHANGES_REQUESTED** [review]
Looks good overall Just a few minor comments:

- Log messages need to be added at a couple of places

- Add comments describing methods and data structure for better code readability

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 2 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: good point, done

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 3 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go:181]: sure, done.

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 4 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go:209]: yep, done

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 5 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go:230]: done!

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 6 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go:247]: done!

File: pkg/topologyinfo/cpus/cpus.go (hunk 7 above)
Comment (ffromani) [reply pkg/topologyinfo/cpus/cpus.go]: done!

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 8 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: fixed

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 9 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: done

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 10 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go:94]: done

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 11 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go:143]: done

File: pkg/topologyinfo/cpus/cpus.go (hunk 1 above)
Comment (ffromani) [reply pkg/topologyinfo/cpus/cpus.go]: fixed!

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+		numaInfo:          &nodes,
+		numaDists:         dists,
+		resourceID2NUMAID: makeResourceMap(len(nodes.Online), allDevs),
+		perNUMACapacity:   MakeNodeCapacity(allDevs),
+	}
+
+	return &nodeResourceInstance, nil
+}
+
+// Aggregate prodices the mapping (numa zone name) -> Zone from the given PodResources.
```
Comment (swatisehgal) [inline pkg/resourcemonitor/noderesourcesaggregator.go]: typo: provides

File: pkg/resourcemonitor/noderesourcesaggregator.go [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+)
+
+const (
+	defaultPodResourcesTimeout = 10 * time.Second
+	// obtained these values from node e2e tests : https://github.com/kubernetes/kubernetes/blob/82baa26905c94398a0d19e1b1ecf54eb8acb6029/test/e2e_node/util.go#L70
+)
+
+type nodeResources struct {
+	perNUMACapacity map[int]map[v1.ResourceName]int64
+	// we should use a different map for each resource type: resourceName -> resourceID -> nodeID
```
Comment (ffromani) [inline pkg/resourcemonitor/noderesourcesaggregator.go]: stale comment, should read like
```
mapping: resourceName -> resourceID -> nodeID
```

File: pkg/topologyinfo/cpus/cpus.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ */
+
+package cpus
```
Comment (swatisehgal) [inline pkg/topologyinfo/cpus/cpus.go]: I noticed [here](https://github.com/fromanirh/topologyinfo/tree/main/cpus) that you already worked on cpu_test.go. If the implementation is exactly the same, can we add the cpu_test.go as part of this PR too? Just a suggestion/reminder :)

**Review by swatisehgal — COMMENTED** [review]
A small typo but other than that we are good to go!

File: pkg/topologyinfo/cpus/cpus.go (hunk 14 above)
Comment (ffromani) [reply pkg/topologyinfo/cpus/cpus.go]: This is exactly where I am expecting the discussion to happen. `topologyinfo` here is indeed a narrowed-down, copy/pasted (poor's man vendor) clone of https://github.com/fromanirh/topologyinfo/. The problem is exactly this: being a copy/paste we do have a test suite, but it is not integrated at all with the rest of the tests found elsewhere in the codebase.

I have no preference regarding how to address this issue (if at all)

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 12 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: thanks, fixed

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 13 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: fixed

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 12 above)
Comment (swatisehgal) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: prodices :) ?

File: pkg/topologyinfo/cpus/cpus.go (hunk 14 above)
Comment (swatisehgal) [reply pkg/topologyinfo/cpus/cpus.go]: Keep me honest, but I thought that [this](https://github.com/swatisehgal/node-feature-discovery/blob/master/Makefile#L103) would take care of executing the tests?

File: pkg/resourcemonitor/noderesourcesaggregator.go (hunk 12 above)
Comment (ffromani) [reply pkg/resourcemonitor/noderesourcesaggregator.go]: ops :) fixing in the next upload

File: pkg/topologyinfo/cpus/cpus.go (hunk 14 above)
Comment (ffromani) [reply pkg/topologyinfo/cpus/cpus.go]: absolutely. But please compare [this](https://github.com/swatisehgal/node-feature-discovery/blob/rte/pkg/topologyinfo/numa/numa_test.go) with [this](https://github.com/swatisehgal/node-feature-discovery/blob/rte/pkg/nfd-worker/nfd-worker-internal_test.go).

We can clearly see the tests I added on `topologyinfo` are written with different style, conventions, tools (`sysfs/fake` instead of `mock`s). This is indeed caused by the poor man's vendoring approach I mentioned. Nevertheless `pkg/topologyinfo` feels alien compared to the rest of the codebase.

Do we want to take care of this? and if so, how?

Comment (swatisehgal) [reply pkg/topologyinfo/cpus/cpus.go]: I see your point. I am open to suggestions on the best way to move ahead. I see a couple of options here:

1. Add the topologyinfo pkg code in this repo for now (merge this PR) and later add tests that are of the same style as that in the repo (as another PR).  

1. Import pkg [https://github.com/fromanirh/topologyinfo](https://github.com/fromanirh/topologyinfo) now, later update the import path to https://github.com/openshift-kni/topologyinfo

Comment (ffromani) [reply pkg/topologyinfo/cpus/cpus.go]: I think the latter is better because makes easier for us to pivot to other information-gathering packages (e.g. `ghw`) later on if we feel it's better or if NFD maintainers ask for it. Updating accordingly.

**Comment by swatisehgal** [comment]
LGTM! Thanks for your great work on this PR :)
